Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,1391 @@
+      }
+
+      func isIdentityOrCompatibleEncoding(_ clientEncoding: CompressionAlgorithm) -> Bool {
+        clientEncoding == .identity || configuration.acceptedEncodings.contains(clientEncoding)
+      }
+
+      // Firstly, find out if we support the client's chosen encoding, and reject
+      // the RPC if we don't.
+      let inboundEncoding: CompressionAlgorithm
+      let encodingValues = metadata.values(
+        forHeader: GRPCHTTP2Keys.encoding.rawValue,
+        canonicalForm: true
+      )
+      var encodingValuesIterator = encodingValues.makeIterator()
+      if let rawEncoding = encodingValuesIterator.next() {
+        guard encodingValuesIterator.next() == nil else {
+          let status = Status(
+            code: .internalError,
+            message: "\(GRPCHTTP2Keys.encoding) must contain no more than one value."
+          )
+          let trailers = self.makeTrailers(status: status, customMetadata: nil, trailersOnly: true)
+          return .rejectRPC(trailers: trailers)
+        }
+
+        guard let clientEncoding = CompressionAlgorithm(rawValue: String(rawEncoding)),
+          isIdentityOrCompatibleEncoding(clientEncoding)
+        else {
+          if configuration.acceptedEncodings.isEmpty {
+            let status = Status(code: .unimplemented, message: "Compression is not supported")
+            let trailers = self.makeTrailers(
+              status: status,
+              customMetadata: nil,
+              trailersOnly: true
+            )
+            return .rejectRPC(trailers: trailers)
+          } else {
+            let status = Status(
+              code: .unimplemented,
+              message: """
+                \(rawEncoding) compression is not supported; \
+                supported algorithms are listed in grpc-accept-encoding
+                """
+            )
+            var customTrailers = Metadata()
+            for acceptedEncoding in configuration.acceptedEncodings {
+              customTrailers.addString(
+                acceptedEncoding.name,
+                forKey: GRPCHTTP2Keys.acceptEncoding.rawValue
+              )
+            }
+            let trailers = self.makeTrailers(
+              status: status,
+              customMetadata: customTrailers,
+              trailersOnly: true
+            )
+            return .rejectRPC(trailers: trailers)
+          }
+        }
+
+        // Server supports client's encoding.
+        inboundEncoding = clientEncoding
+      } else {
+        inboundEncoding = .identity
+      }
+
+      // Secondly, find a compatible encoding the server can use to compress outbound messages,
+      // based on the encodings the client has advertised.
+      let outboundEncoding: CompressionAlgorithm
+      let clientAdvertisedEncodings = metadata.values(
+        forHeader: GRPCHTTP2Keys.acceptEncoding.rawValue,
+        canonicalForm: true
+      )
+      .compactMap { CompressionAlgorithm(rawValue: String($0)) }
```

Comment:
It should be possible to avoid this array here by doing a plain for loop. We need to be very careful with allocations since we'll do this for every request.